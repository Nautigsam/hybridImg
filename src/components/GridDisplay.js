import React, { useState } from 'react';
import {
  Table,
  Notification,
  Button,
  Form,
  Level,
  Icon,
} from 'react-bulma-components';

import GridCell from './GridCell';
import StoreProvider from '../store/StoreProvider';
import useStateWithLocalStorage from './useStateWithLocalStorage';
import { updateCell } from '../store/StoreActions';

export const GridDisplay = ({
  grid,
  users,
  cells,
  getCellCandidates,
  dispatchToStore,
}) => {
  const [filter, setFilter] = useStateWithLocalStorage('filter', {});
  const [loading, setLoading] = useState(false);

  const autoFill = async () => {
    setLoading(true);
    Promise.all(
      cells.map(cell => {
        const candidates = getCellCandidates(cell);
        if (candidates.length > 0) {
          return dispatchToStore(
            updateCell({
              id: cell.id,
              hybrids: [...cell.hybrids, ...candidates],
            })
          );
        }
      })
    ).then(() => {
      setLoading(false);
    });
  };

  return (
    <>
      <Notification color="primary" className="is-light">
        <Level>
          <Level.Side>
            <Level.Item>
              <Form.Label>Filter par auteurice :</Form.Label>
            </Level.Item>
            <Level.Item>
              <Form.Select
                value={filter.user}
                onChange={e => setFilter({ ...filter, user: e.target.value })}
              >
                <option></option>
                {users.map(u => (
                  <option key={u.id} value={u.id}>
                    {u.name}
                  </option>
                ))}
              </Form.Select>
            </Level.Item>
          </Level.Side>
          <Level.Side>
            <Level.Item>
              <Button color="primary" onClick={autoFill}>
                Remplir automatiquement
                {loading ? (
                  <span className="is-primary button is-loading"></span>
                ) : (
                  <></>
                )}
              </Button>
            </Level.Item>
          </Level.Side>
        </Level>
      </Notification>
      <Table>
        <thead>
          <tr>
            <th></th>
            {grid.colThemes.map(col => (
              <th key={col.id}>{col.name}</th>
            ))}
          </tr>
        </thead>
        <tbody>
          {grid.lineThemes.map(line => (
            <tr key={line.id}>
              <th>{line.name}</th>
              {grid.colThemes.map(col => (
                <GridCell
                  key={col.id}
                  grid={grid.id}
                  line={line}
                  col={col}
                  filter={filter}
                />
              ))}
            </tr>
          ))}
        </tbody>
      </Table>
    </>
  );
};

function extraProps(store, props) {
  const grid = store.getGrid(props.match.params.id);
  return {
    grid,
    cells: Object.values(store.cells).filter(c => c.grid === grid.id),
    getCellCandidates: cell => {
      return Object.values(store.hybrids)
        .filter(
          h =>
            h.user === store.session.user.id &&
            h.tags.includes(cell.position.col) &&
            h.tags.includes(cell.position.line) &&
            !cell.hybrids.includes(h.id)
        )
        .map(h => h.id);
    },
    users: store.getGridUsers(grid.id),
  };
}

export default StoreProvider(extraProps)(GridDisplay);
