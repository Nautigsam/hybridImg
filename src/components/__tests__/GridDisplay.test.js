import React from 'react';
import { shallow } from 'enzyme';

import { GridDisplay } from '../GridDisplay';

describe('GridDisplay', () => {
  it('renders correctly', () => {
    const props = {
      grid: {
        id: 'G1',
        name: 'La grille de Mag',
        lineThemes: ['T1', 'T2', 'T3', 'T4', 'T5', 'T6'],
        colThemes: ['T7', 'T8', 'T9', 'T10', 'T11', 'T12'],
      },
      users: [{ id: 'u1', name: 'Mag' }, { id: 'u2', name: 'Matt' }],
      edit: true,
    };
    const wrapper = shallow(<GridDisplay {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
