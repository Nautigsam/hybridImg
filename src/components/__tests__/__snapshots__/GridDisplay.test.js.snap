// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridDisplay renders correctly 1`] = `
<Fragment>
  <Table
    bordered={false}
    size="fullwidth"
    striped={true}
  >
    <thead>
      <tr>
        <th />
        <th />
        <th />
        <th />
        <th />
        <th />
        <th />
      </tr>
    </thead>
    <tbody>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T1"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T1"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T1"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T1"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T1"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T1"
        />
      </tr>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T2"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T2"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T2"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T2"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T2"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T2"
        />
      </tr>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T3"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T3"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T3"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T3"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T3"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T3"
        />
      </tr>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T4"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T4"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T4"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T4"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T4"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T4"
        />
      </tr>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T5"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T5"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T5"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T5"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T5"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T5"
        />
      </tr>
      <tr>
        <th />
        <Memo(HybridCellContainer)
          col="T7"
          edit={true}
          grid="G1"
          line="T6"
        />
        <Memo(HybridCellContainer)
          col="T8"
          edit={true}
          grid="G1"
          line="T6"
        />
        <Memo(HybridCellContainer)
          col="T9"
          edit={true}
          grid="G1"
          line="T6"
        />
        <Memo(HybridCellContainer)
          col="T10"
          edit={true}
          grid="G1"
          line="T6"
        />
        <Memo(HybridCellContainer)
          col="T11"
          edit={true}
          grid="G1"
          line="T6"
        />
        <Memo(HybridCellContainer)
          col="T12"
          edit={true}
          grid="G1"
          line="T6"
        />
      </tr>
    </tbody>
  </Table>
</Fragment>
`;
